.github/actions/build-action/action.yml

```diff
@@ -0,0 +1,162 @@
+outputs:
+  random-number:
+    description: "Random number"
+    value: ${{ steps.random-number-generator.outputs.random-id }}
+runs:
+  using: "composite"
+  steps:
+    - name: Dump-environment
+      shell: bash
+      run: |
+        env
+        set | grep =
+
+    - name: Install ccache
+      shell: bash
+      run: |
+        if ! type -a ccache
+        then
+          apt install ccache
+        fi
+
+    - name: Reset ccache stats
+      shell: bash
+      run: |
+          ccache -s
+          # Reset the stats so only this build counts
+          ccache -z
+
+    - name: Disk space
+      shell: bash
+      run: |
+        df -h
+        if [[ -d "${CACHE_DIR}" ]]
+        then
+            du -sh ${CACHE_DIR}
+        fi
+        if [[ -d "${CCACHE_DIR}" ]]
+        then
+            du -sh ${CCACHE_DIR}
+        fi
+        if [[ -d "${NIH_CACHE_ROOT}" ]]
+        then
+            find ${NIH_CACHE_ROOT} -maxdepth 3 \( -iname src -prune -o \
+              -type d -exec du -sh {} \; \)
+            # Post-condition: The cache will not be reused to build with a
+            # different compiler
+            if [[ "${CC}" =~ "gcc" ]]
+            then
+              echo Erasing NIH cache for clang
+              rm -rf "${NIH_CACHE_ROOT}"/*/Clang_*/ || true
+            elif [[ "${CC}" =~ "clang" ]]
+            then
+              echo Erasing NIH cache for gcc
+              rm -rf "${NIH_CACHE_ROOT}"/*/GNU_*/ || true
+            fi
+            # Free up space if the cache is being reused with a different
+            # generator from prior builds (ninja vs. makefiles).
+            # Post condition: Cache will not be reused with a different
+            # generator.
+            if [[ "${NINJA_BUILD}" == "true" && -d "${NIH_CACHE_ROOT}"/unix_makefiles && \
+              $( du -sm "${NIH_CACHE_ROOT}"/unix_makefiles | cut -f1 ) -gt 1000  ]]
+            then
+              echo Erasing NIH cache for makefiles
+              rm -rf "${NIH_CACHE_ROOT}"/unix_makefiles || true
+            elif [[ "${NINJA_BUILD}" == "false" && -d "${NIH_CACHE_ROOT}"/ninja && \
+              $( du -sm "${NIH_CACHE_ROOT}"/ninja | cut -f1 ) -gt 1000  ]]
+            then
+              echo Erasing NIH cache for ninja
+              rm -rf "${NIH_CACHE_ROOT}"/ninja || true
+            fi
+        fi
+
+    - name: Build and test
+      shell: bash
+      run: |
+          export CXX=${CC/gcc/g++}
+          export CXX=${CXX/clang/clang++}
+          export NUM_PROCESSORS=${NUM_PROCESSORS:-$( nproc )}
+          echo NUM_PROCESSORS is ${NUM_PROCESSORS}
+          if [ "${{ inputs.image }}" == "rippled-ci-builder:2944b78d22db" ]
+          then
+              export CMAKE_EXTRA_ARGS+=" -DBoost_NO_BOOST_CMAKE=ON"
+          fi
+          if [ "${BUILD_TYPE}" == "Release" ]
+          then
+              export CMAKE_EXTRA_ARGS+=" -Dassert=ON"
+          fi
+
+          # Override environment variables that don't work smoothly
+          # using env:
```

Comment:
Environment variables can be specified using `env:` at the workflow, job, and step level. I can't remember if particular comment/construction because the `inputs` object was not available at that higher job level, or if defining it there would set some of those variables to be empty (instead of leaving them unset), and causing undesirable side effects, but I'm pretty sure it was one of those two.

Reply:
I tried following some of these variables. It seems `NINJA_BUILD` is only used in `bin/ci/ubuntu/build-and-test.sh`, and can be set to an empty value, except that this use probably needs to be corrected to `${NINJA_BUILD:-false}`:

https://github.com/ripple/rippled/blob/10e2c6d66c9ce7e530269e15684ca3fbbb88fc42/bin/ci/ubuntu/build-and-test.sh#L67

I believe all the `*SAN_OPTIONS` variables can be empty. I tested `UBSAN_OPTIONS` on a build with `-Dsan=undefined` and it worked.

I'd like to shrink this section as small as possible. Programs should be robust in the presence of empty variable values.

Going down this path brought my attention to a bigger issue. We've got files under `bin/ci/ubuntu` that we expect to run on Windows. It was likely written first with Ubuntu in mind, and likely not robustly enough to handle empty strings for some environment variables. I've already tried [once](https://github.com/ripple/rippled/commit/c6d82c722b0d2bc66ce9c40f0e0158ac84176b75) to fix this problem, when I wrote the pipelines for GitLab. Can we build on that effort and:

- redesign the scripts under `bin/ci/ubuntu` to handle empty environment variables correctly
- remove any unused scripts (I haven't yet checked whether each is used)
- document them (I wrote a README for my scripts, and left helpful comments at the top of each documenting their interfaces), and
- remove "ubuntu" from their path if they are cross-platform?

Reply:
There's a lot here, and I haven't addressed all of it, but one simple thing I did was to change the `== false` comparisons to `!= true`, which will make them more robust in the face of empty values and such.